File: benchmarks-haskell/ackermann-of-3-3.hs
import Interpreter

isZero :: Term
isZero = UnaryOp (\x -> if x == 0 then 1 else 0)

incr :: Term
incr = UnaryOp (+ 1)

decr :: Term
decr = UnaryOp (\x -> x - 1)

ackermann :: Term
ackermann =
  Fix
    ( Lambda
        ( Lambda
            ( Lambda
                ( IfThenElse
                    (Apply isZero (TVar 1))
                    (Apply incr (TVar 0))
                    ( IfThenElse
                        (Apply isZero (TVar 0))
                        (Apply (Apply (TVar 2) (Apply decr (TVar 1))) (Const 1))
                        ( Apply
                            (Apply (TVar 2) (Apply decr (TVar 1)))
                            (Apply (Apply (TVar 2) (TVar 1)) (Apply decr (TVar 0)))
                        )
                    )
                )
            )
        )
    )

benchmarkTerm :: Term
benchmarkTerm = Apply (Apply ackermann (Const 3)) (Const 3)

main :: IO ()
main = case denote [] benchmarkTerm of
  VConst n -> print n
  _ -> error "Expected a constant result!"

File: benchmarks-ocaml/ackermann_of_3_3.ml
open Interpreter

let is_zero = UnaryOp (fun x -> if x = 0 then 1 else 0)

let incr = UnaryOp (( + ) 1)

let decr = UnaryOp (fun x -> x - 1)

let ackermann =
  Fix
    (Lambda
       (Lambda
          (Lambda
             (IfThenElse
                ( Apply (is_zero, tvar 1),
                  Apply (incr, tvar 0),
                  IfThenElse
                    ( Apply (is_zero, tvar 0),
                      Apply (Apply (tvar 2, Apply (decr, tvar 1)), Const 1),
                      Apply
                        ( Apply (tvar 2, Apply (decr, tvar 1)),
                          Apply (Apply (tvar 2, tvar 1), Apply (decr, tvar 0))
                        ) ) )))))

let benchmark_term = Apply (Apply (ackermann, Const 3), Const 3)

let () =
  match denote [] benchmark_term with
  | VConst n -> Printf.printf "%d\n" n
  | _ -> failwith "Expected a constant result!"
